From 6b0f84fe705746eebb3195bc5858606b0be9ea03 Mon Sep 17 00:00:00 2001
From: icefairy64 <fake@mail.local>
Date: Wed, 29 May 2024 19:34:13 +0400
Subject: [PATCH] Model Tilt > initial commit

---
 comfy/model_patcher.py | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/comfy/model_patcher.py b/comfy/model_patcher.py
index 2e746d8..e7f6977 100644
--- a/comfy/model_patcher.py
+++ b/comfy/model_patcher.py
@@ -487,6 +487,18 @@ class ModelPatcher:
                         weight += ((strength * alpha) * lora_diff).type(weight.dtype)
                 except Exception as e:
                     logging.error("ERROR {} {} {}".format(patch_type, key, e))
+            elif patch_type == "tilt":
+                tp = weight.dtype
+
+                gen = torch.manual_seed(v[0])
+                noise = torch.randn(size=weight.size(), dtype=tp, layout=weight.layout, generator=gen, device='cpu')
+                noise = (noise.to(weight.device) - 0.5) * strength * 2
+
+                gen = torch.manual_seed(v[2])
+                mask_noise = torch.randn(size=weight.size(), dtype=tp, layout=weight.layout, generator=gen, device='cpu')
+                noise *= ((v[1] - mask_noise.to(weight.device)) * 1000000).clamp(0, 1)
+
+                weight += noise
             else:
                 logging.warning("patch type not recognized {} {}".format(patch_type, key))
 
-- 
2.45.1

